core/src/main/java/io/questdb/griffin/engine/functions/regex/AbstractLikeVarcharFunctionFactory.java

```diff
@@ -144,13 +145,117 @@ public Function newInstance(
 
     protected abstract boolean isCaseInsensitive();
 
+    /**
+     * Optimized variant of {@link ConstContainsVarcharFunction} with SWAR-based contains implementation.
+     * Works only for patterns up to 8 bytes in size.
+     */
+    private static class ConstContainsSwarVarcharFunction extends BooleanFunction implements UnaryFunction {
+        private static final int MAX_SIZE = Long.BYTES;
+        private final Utf8Sequence pattern; // only used in toPlan
+        private final long patternMask;
+        private final int patternSize;
+        private final long patternWord;
+        private final byte searchFirstByte;
+        private final long searchWord;
+        private final Function value;
+
+        public ConstContainsSwarVarcharFunction(Function value, Utf8Sequence pattern) {
+            assert pattern.size() > 0 && pattern.size() <= MAX_SIZE;
+            this.value = value;
+            this.patternSize = pattern.size();
+            this.patternMask = patternSize == MAX_SIZE ? -1L : (1L << 8 * pattern.size()) - 1L;
+            long patternWord = 0;
+            for (int i = 0, n = pattern.size(); i < n; i++) {
+                patternWord |= (long) (pattern.byteAt(i) & 0xff) << (8 * i);
+            }
+            this.patternWord = patternWord;
+            this.searchFirstByte = pattern.byteAt(0);
+            this.searchWord = SwarUtils.broadcast(pattern.byteAt(0));
+            this.pattern = pattern;
+        }
+
+        @Override
+        public Function getArg() {
+            return value;
+        }
+
+        @Override
+        public boolean getBool(Record rec) {
+            Utf8Sequence us = value.getVarcharA(rec);
+            if (us == null || us.size() < patternSize) {
+                return false;
+            }
+
+            int size = us.size();
+
+            int i = 0;
+            for (int n = size - 7 - patternSize + 1; i < n; i += 8) {
```

Comment:
@mtopolnik I've noticed a bug in this loop when merging with `master`. Namely, `n` was `size - 7` while we were looking ahead by potentially 14 bytes (`us.longAt(i)` to match the first byte and then `us.longAt(i + foundIndex)` in a `while (foundIndex < 8)` loop). This could have led to a segfault. So, I fixed the bug by fixing the loop boundary and adding a check if it's safe to load full word to do the pattern match check.

Besides that, I've moved `Long.numberOfTrailingZeros()` call into `SwarUtils#indexOfFirstMarkedByte()` since it's also needed in the CSV parser for similar purposes.

Could you check the changes?

Reply:
I'd write

```java
for (int n = size - 15; i < n; i += Long.BYTES) {
```

This would mean:

```plain
max(i) = size - 14
max(pos) = 7
max(i + pos) = size - 14 + 7 = size - 7
```

and `size - 7` is exactly the max offset at which `longAt` is safe.

Then we can eliminate the check `if (size - i - pos > 7)`, and don't need the special-cased tail code within the non-tail loop.

Reply:
I've already tried that approach and it appears to be slower than the current one in clickbench:
```sql
-- current loop: 985ms
-- suggested loop: 1.1s
SELECT COUNT(*) FROM hits WHERE URL LIKE '%google%';
```

Reply:
Looks like one extra first byte check makes difference for shorter strings.